ansible_base/lib/routers/association_resource_router.py

```diff
@@ -0,0 +1,187 @@
+        manager.add(*related_instances)
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    @action(detail=False, methods=['post'])
+    def disassociate(self, request, **kwargs):
+        """
+        Disassociate a related object from this object.
+
+        This will be served at /{basename}/{pk}/{related_name}/disassociate/
+        We will be given a list of primary keys in the request body.
+        """
+        instance = self.backward_related_qs.get(pk=kwargs['pk'])
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        related_instances = serializer.validated_data['instances']
+        if not related_instances:
+            raise serializers.ValidationError({'instances': _('Please pass in one or more instances to disassociate')})
+        manager = getattr(instance, self.association_fk)
+
+        # Ensure each of the given related_instances is actually related to the instance.
+        # If any isn't, then bomb out and tell the user which ones aren't related.
+        given_related_instance_set = set(related_instances)
+        related_instance_set = set(manager.all())
+        non_related_instances = given_related_instance_set - related_instance_set
+        if non_related_instances:
+            raise serializers.ValidationError(
+                {
+                    'instances': _('Cannot disassociate these objects because they are not all related to this object: %(non_related_instances)s')
+                    % {'non_related_instances': ', '.join(str(i.pk) for i in non_related_instances)},
+                }
+            )
+
+        try:
+            manager.remove(*related_instances)
+        except AttributeError:
+            # If the field is FK and non-nullable, .remove() won't exist.
+            # https://docs.djangoproject.com/en/dev/ref/models/relations/#django.db.models.fields.related.RelatedManager.remove
+            # "For ForeignKey objects, this method only exists if null=True."
+            # Return something stating that the operation is not supported.
+            raise serializers.ValidationError(
+                {
+                    'instances': _('Cannot disassociate these objects because there must be a related object'),
+                }
+            )
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    def get_serializer_class(self):
+        association_class = getattr(self, 'association_serializer', None)
+        if self.action in ('associate', 'disassociate'):
+            if association_class is None:
+                raise RuntimeError('You must set association_serializer on the viewset')
+            return association_class
+        return super().get_serializer_class()
+
+
+class ReverseViewMixin(QuerySetMixinBase):
+    def get_serializer_class(self):
+        related_class = getattr(self, 'related_serializer', None)
+        if self.action in ('associate', 'disassociate'):
```

Comment:
This feels weird to me. These actions are defined in `AssociateMixin` which (in our code that subclasses the mixins) is mutually exclusive with `ReverseViewMixin` right?

So `self.action` will never be these actions because these actions (presumably) don't exist.

Reply:
Fixed... ish. Has a bug that I need to track down but it now it uses class property.